### ISSUE TITLE
VQGAN generated codes mismatch with given data examples

### ISSUE THREAD

**Issue by Lizw14** [OFA-Sys/OFA#258]
I am preparing data to pretrain using only CC12M dataset. 

When preparing the input .tsv data file, I generate the VQ-GAN codes for the infilling task using [run_scripts/image_gen/generate_for_vqgan_code.sh](https://github.com/OFA-Sys/OFA/blob/main/run_scripts/image_gen/generate_for_vqgan_code.sh). However, the generated codes do not match with ones provided in the pretrain_data_examples.zip files as provided in the README. (I am sure that I use exactly the same images - modified from the given data examples, sent for VQGAN code generation, then compare back with the code in the given data examples.)

The codes generated by generate_for_vqgan_code.sh is of length 1024, while the length in the data examples is 256. I notice that in the README, it is said that "generate the sparse codes for **the central part of image** through VQ-GAN". I don't know if ``CenterCrop`` could be a possible reason for the difference. Could you help look into this and suggest which version to use?

Additionally, I find that the result of [run_scripts/image_gen/generate_for_vqgan_code.sh](https://github.com/OFA-Sys/OFA/blob/main/run_scripts/image_gen/generate_for_vqgan_code.sh) differs in multiple runs. I run the same scripts with the same inputs for multiple times, and their outputs are different (but similar on some positions). Are there anything random in the code generation process?

Thanks in advance and I appreciate the solid work!

**Comment by Lizw14** [comment]
Additionally, I find that the result of [run_scripts/image_gen/generate_for_vqgan_code.sh](https://github.com/OFA-Sys/OFA/blob/main/run_scripts/image_gen/generate_for_vqgan_code.sh) differs in multiple runs. I run the same scripts with the same inputs for multiple times, and their outputs are different (but similar on some positions). Are there anything random in the code generation process?

**Comment by jxst539246** [comment]
1. During pretraining, we use the image infilling task, in which we use the masked image as input and generate the central part of image, so during pretraining, the length of image code is 256 (for the central part of image 128 * 128). During fine-tuning, we use the image generation task, in which we will generate the whole 256*256 image (so the length of image code becomes 1024).

2. You are right, as the vqgan model we use contains a GumbelQuantize, the gumbel_softmax function will randomly sample from the logits. For the same inputs images, it will encode them into different (but similar) image code sequences. If you fed these image code sequences back into vqgan and decode them to images, you will find that the decoded images are almost the same.

hope these comments can solve your questions.

**Comment by Lizw14** [comment]
Great, this solves my question. Thanks for the answer!

**Comment by varadgunjal** [comment]
@Lizw14 @jxst539246 Can you help me understand how this 256-long image codes in `image_examples.tsv` are generated for pretraining purposes? Do we just generate the 1024 long code and pick out the 256 corresponding to the central patch of the original image?
